### ISSUE TITLE
Feature request: Make packages/modules more program-like

### ISSUE THREAD

**Issue by CMDJojo** [modmark-org/modmark#38]
**Is your feature request related to a problem? Please describe.**
This feature request is related to the API implemented by packages/modules and accessed by `core` when loading a module.

A `package` needs to declare the following information: 
 - `name` (the name of the package)
 - `version` (the version of the package)
 - `transforms` (the module invocations/node transformations the package handles, including arguments for the case of module invocations)
Additionally, the `package` needs a way to actually execute the transforms.

Currently, the information declarations are implemented as exposed library functions. This means that a package needs to be written as a library with exposed public un-mangled functions. For rust, this thus needs `#[no_mangle]` for each function, and if written in another language, care needs to be taken to not mangle it (`extern "c"` in C++, `_cdecl()` in Swift etc). The function transformation execution is currently implemented by executing the entry point of the package, which also means that it must be written as an executable, and have an actual entry point, be able to read arguments, stdin, stdout etc.

I think this is problematic since you need to write the module as both an executable and a library at the same time, of sorts, and it adds a lot of boiler plate just for that. Here is an example of a compliant module written this way (with the transforms unimplemented):
```rust
use std::{env, io};

#[no_mangle]
pub fn name() {
    println!("Module test");
}

#[no_mangle]
pub fn version() {
    println!("1");
}

#[no_mangle]
pub fn transforms() {
    println!("[table] -> table");
    println!("border = black - What color the border should be");
    println!("");

    println!("table -> html latex");
    println!("");

    println!("row -> html latex");
    println!("");
}

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();
    let node_name = &args[0];

    let mut body = String::new();
    io::stdin().read_line(&mut body).unwrap();

    match node_name.as_str() {
        "[table]" => todo!(),
        "row" => todo!(),
        "table" => todo!(),
        other => eprintln!("This node does not support node conversions from {other}"),
    }
}
```
Additionally, it is problematic since the intuition that "a module acts just like a program ran from the terminal, it reads arguments, stdin, stdout etc" isn't really upheld here; a normal terminal application doesn't also expose library functions.

**Describe the solution you'd like**
I would like the modules to act more like a simple terminal tool, which means just one entry point. To me, it seems reasonable to bundle all required information `name`, `version`, `transforms` into one `manifest`. The first argument to the tool may be either `manifest` to get the manifest, or `invoke` to handle an invocation, or `transform` to handle a transform (or maybe just one argument combining `invoke` and `transform` if they are implemented in the exact same way, the reason I chose to make them different is because `invoke` declares arguments that are required when calling whereas node transform currently doens't). For the `manifest`, no additional arguments will be provided, and for `invoke`/`transform`, the arguments now given are following (as for choosing what transform to execute).

The manifest for the example above might look something like this:
```
Module test@v1
[table] -> table
border = black - What color the border should be

table -> html latex

row -> html latex
```

The entire module might be declared like this:
```rust
use std::{env, io};
let manifest = r"
Module test@v1
[table] -> table
border = black - What color the border should be

table -> html latex

row -> html latex"

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();
    let action = &args[0]
    match action.as_str() {
        "manifest" => print!(manifest)
        "invoke" => invoke(&args[1], &args[2..])
        "transform" => transform(&args[1])
    }
}

fn invoke(module: &str, args: &[String]) {
    if module == "table" {
        todo!()
    } else {
        eprintln!("No such module invocation [{}]", module)
    }
}

fn transform(node: &str) {
    match node {
        "row" => todo!(),
        "table" => todo!(),
        other => eprintln!("No such transform from {other}")
    }
}
```
I think this improves readability and comprehension of the actual execution of the module. It also makes it act more like a normal terminal tool, where `tool-name help` is a common usage, whereas we have `package manifest` to get the manifest. One additional benefit is that `manifest` is a recognized term; many plugins for browsers needs to supply a `manifest.json` declaring their intents, privileges and accesses. If later revisions also made packages declare what environment variables they require, access, and modify, it would be suitable to be added at the end of the manifest. I think this change also will make it easier to make libraries for declaring packages etc.

**Describe alternatives you've considered**
I don't think it is good to mix library and executables as we currently do. One alternative might be to just make it as an library, and have our transformation declared as a library function, but I think it makes more sense to see it as an executable.

The manifest syntax might be changed to a format easier to parse (maybe JSON), this feature request only supplies one example of how a manifest can be derived from our current example. For example, the version might be split to a separate line, and markers might be added to define different sections, one for general information about the package, one for the transforms etc. It might be suitable with toml-like syntax for those, in that case.

One alternative to doing `package manifest` to get all information might be to have multiple different commands, `package name` to get the name, `package transforms` to get the transforms etc. This will make it more like the current implementation, with all information-containing sections divided into different commands. However, I prefer the `package manifest` way since we only need to call the package once to get all required information as opposed to one call per piece of information. This will reduce wasm overhead as well as the potential overhead by the module (if it has some start-up time).

**Comment by axoxelol1** [comment]
Seems like a reasonable request, would probably make the packages easier to test using the command line as well. Should not be a difficult change to make either.

**Comment by adelhult** [comment]
This is fixed in #64.
